### DISCUSSION TITLE
RuntimeError: More than one spatial_shape found in subject images

### DISCUSSION THREAD

**Discussion by ruby0101** [TorchIO-project/torchio#489] [General]
Hi, I am trying to train U-Net model for tumor segmentation with my MRI T1 image dataset. I just followed the tutorial 
But when I call a train function, it gave me the following error:

```python-traceback
RuntimeError: More than one spatial_shape found in subject images: {'brain': (76, 76, 48), 'mri': (75, 75, 48)}
```

I have already applied training_transform which is in the tutorial  `tio.CropOrPad((48, 60, 48))` to the training and validation dataset. But the error message, it seems like transformation didn't work properly. The processed sample size is `{'brain': (76, 76, 48), 'mri': (75, 75, 48)}`. Could you help me to solve it, please? What is a problem with it?

**Comment by fepegar** [comment]
Hi, @ruby0101. It's hard to diagnose without any code or data, but my guess is your `brain` and `mri` images have different associated affine matrices. Do you think you can provide a minimally working example as asked in the [bug report template](https://github.com/fepegar/torchio/issues/new?assignees=&labels=bug&template=bug_report.md&title=)?

**Reply by ruby0101** [reply]
Thank you so much for your prompt response. Could you check my code please? I uploaded it to my google drive. https://drive.google.com/file/d/1sHzZ6rh54r3Z-Ck0ksWX1KdgaPSIvbkI/view?usp=sharing
I think transformation has not applied to all images.

**Reply by fepegar** [reply]
Can you please upload that to Google Colab so I can run it remotely?

**Reply by ruby0101** [reply]
Sorry, I don't have a permission to upload the dataset into google drive. Could you check my runned jupyter notebook file? Nothing is different from tutorial code. Only dataset is different. I added some print code for debugging. If you see the print result, some image file wasn't transformed properly.

**Reply by fepegar** [reply]
I just meant that you can upload the notebook to Google Colab, like this: https://colab.research.google.com/drive/1sHzZ6rh54r3Z-Ck0ksWX1KdgaPSIvbkI

As I said, the spatial metadata is probably not consistent between your MRI and the corresponding segmentation for the subject that provokes the error. I believe you can check that with `assert np.all(np.equal(subject.mri.affine, subject.brain.affine))`.

**Reply by ruby0101** [reply]
Hi, I printed out all code steps one by one as debugging purpose. Finally I have found the transformation contains a bug. I mean this: "training_set = tio.SubjectsDataset(training_subjects, transform=training_transform)"
Please see the code for details:
https://colab.research.google.com/drive/1Z0Dw9xr_35irhxiQbQ6oU7wdNa8RnGf0?usp=sharing

**Reply by fepegar** [reply]
Again: it looks like your images have different metadata (`affine` attribute of `Image` class). That would cause images within the same subject to have different shapes after `Resample`, which would, in turn, generate an error in `CropOrPad`.

```python
In [1]: import torchio as tio

In [2]: import numpy as np

In [3]: affine_1 = np.eye(4)

In [4]: affine_2 = affine_1.copy()

In [5]: affine_2[0, 0] *= 1.01

In [6]: affine_1
Out[6]:
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [7]: affine_2
Out[7]:
array([[1.01, 0.  , 0.  , 0.  ],
       [0.  , 1.  , 0.  , 0.  ],
       [0.  , 0.  , 1.  , 0.  ],
       [0.  , 0.  , 0.  , 1.  ]])

In [8]: import torch

In [12]: transform = tio.Resample(4)

In [15]: im1 = tio.ScalarImage(tensor=torch.rand(1, 256, 256, 192), affine=affine_1)

In [16]: im2 = tio.ScalarImage(tensor=torch.rand(1, 256, 256, 192), affine=affine_2)

In [17]: subject = tio.Subject(im1=im1, im2=im2)

In [18]: transformed = transform(subject)

In [19]: transformed.shape
```

```python-traceback
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-19-30414fc1adec> in <module>
----> 1 transformed.shape

~/git/torchio/torchio/data/subject.py in shape(self)
     91         Consistency of shapes across images in the subject is checked first.
     92         """
---> 93         self.check_consistent_attribute('shape')
     94         return self.get_first_image().shape
     95

~/git/torchio/torchio/data/subject.py in check_consistent_attribute(self, attribute)
    205                 f'\n{pprint.pformat(values_dict)}'
    206             )
--> 207             raise RuntimeError(message)
    208
    209     def check_consistent_spatial_shape(self) -> None:

RuntimeError: More than one shape found in subject images:
{'im1': (1, 64, 64, 48), 'im2': (1, 65, 64, 48)}
```

**Reply by ruby0101** [reply]
Thanks a lot, you are right, I have followed your suggestion and checked it, affines are different. Please see the screenshot (https://drive.google.com/file/d/1DIYph88HDfivgai9dlR6Mor-kku7u1DD/view?usp=sharing). Is there any better way (direct available function) to make affines the same for each image of whole dataset? (instead of updating the each image)

**Reply by fepegar** [reply]
You could resample all your label maps using their corresponding MRI as reference before training. Something like this:

```python
transform = tio.Resample('mri')
for subject in subjects:
    transformed = transform(subject)  # this ensures that subject.brain.affine is equal to subject.mri.affine
    transformed.brain.save(new_seg_path)
```

**Reply by ruby0101** [reply]
Thank you so much. Sorry, I am a fresh on the medical imaging. Could you tell me please how can I resample all the label maps using the corresponding MRI? I mean how could I get new_seg_path? I have tried the below several ways to update affine array. But actually it didn't work. We can't update affine like this, right? Is that possible to update affine metadata directly? such as rounding or truncating the values?
1. dataset[0].brain.affine[1][3] = dataset[0].mri.affine[1][3]
2. for i in range(1,len(dataset)):
       dataset[i].mri.affine = np.round(dataset[i].mri.affine, decimals=3)
       dataset[i].brain.affine = np.round(dataset[i].brain.affine, decimals=3)

**Reply by ruby0101** [reply]
Is there any torchio function like nibabel functions such as set_sform? / e.g. n1_header.set_sform(np.diag([3, 4, 5, 1]), code='mni')/

**Reply by fepegar** [reply]
`new_seg_path` is just the variable you set to choose a new filepath for your label images. For example, `seg_path = 'subject_1_seg.nii.gz'` and `new_seg_path = 'subject_1_seg_mri_affine.nii.gz'`. Or you could just overwrite your label images. You could even do this in bash using `torchio-transform`.

**Reply by ruby0101** [reply]
I see, do you mean that I can update label images directly with its image data? For example: my image file is mri00001.nii.gz and label image is label00001.nii.gz. Their affines are different. Thus could I overwrite the label image directly like the following? Sorry if I misunderstood it. 
    transformed.brain.save('mri00001.nii.gz')

**Reply by fepegar** [reply]
Yes. Although I'm not sure what `transformed` is in your code. Or you could put `Resample('mri')` at the beginning of your composed transforms, in case you want to keep your original images.

**Reply by ruby0101** [reply]
Thanks, but I have a memory issue when I run the code. Please check the screenshot: https://drive.google.com/file/d/1jpaPE5JLuaOgnO2dHA4xO1G5dbQs5XH-/view?usp=sharing
I don't know why my memory gets full easily just running this piece of code. I cleaned my garbage and deleted the used variable in each iteration. But memory is not released just filling up to 100% then my laptop gets on stuck.  Actually I have 32GB memory. What's wrong with this code? Which line of code fills up my memory? Could you help me to figure it up?

**Reply by fepegar** [reply]
Sure! This happens because you are storing all the data inside each subject included in `fixed_subjects` and in `subjects`.

Remember TorchIO uses lazy loading, i.e., voxel data is loaded when needed and then cached in `image.data`. In your code, you are storing data in each subject in `subjects` when you apply the transform, and the transformed data is also stored in each `fixed` subject.

My suggestion:

```python
dataset = tio.SubjectsDataset(subjects, transform=transform)
for image_path, transformed_subject in zip(image_paths, dataset):
    transformed_subject.brain.save(image_path)
```

Then you can create a new list of subjects from `image_paths` for your training dataset.

**Reply by ruby0101** [reply]
transformed_subject.brain.save(image_path), oh, save function changed the original data file directly instead of virtual data, right? after I run the code, I restarted my kernel and run my code again. But a sample plotting now only shows the label. Before I run the save function it was like this: https://drive.google.com/file/d/1oCefNBdAlKX9M0izEaqPbTfuTy0-BBq-/view?usp=sharing
But now it is https://drive.google.com/file/d/1Rc6_-VDmW2IfBg24uEHt0tSp50F4PeRl/view?usp=sharing.

**Reply by ruby0101** [reply]
Many thanks for your prompt reply and advices. Could you check my code please? https://drive.google.com/file/d/1lMtUpYE_VI5qrW-29qXrBGsYJvuTakS9/view?usp=sharing
I am just wondered how do I feed original image data for training dataset after I updated it with label data. Could you clarify me on this please?

**Reply by fepegar** [reply]
> Could you check my code please?

I would appreciate it if you could paste syntax-highlighted code and upload images here, on GitHub, so I don't have to go to all those links.

> I am just wondered how do I feed original image data for training dataset after I updated it with label data. Could you clarify me on this please?

I am not sure what you mean, sorry. Could you elaborate?

**Reply by ruby0101** [reply]
Thanks a lot, finally I have solved the affine issues.

**Reply by ruby0101** [reply]
Model training part, another issue is raised. When training the model, it gave the below error. 
<img width="1066" alt="Screen Shot 2021-03-28 at 12 35 38 AM" src="https://user-images.githubusercontent.com/72773297/112727491-ab15df80-8f5d-11eb-96e2-42244f36645d.png">
I checked the tutorial notebook file, torch sizes are as same as mine. 
<img width="603" alt="Screen Shot 2021-03-28 at 12 40 20 AM" src="https://user-images.githubusercontent.com/72773297/112727590-24153700-8f5e-11eb-9802-79ca49eade9e.png">
Could you check it please? why it gives me the torch size error in my code?

**Reply by fepegar** [reply]
How do you instantiate the data loader?

**Reply by ruby0101** [reply]
Finally, it only worked with training_batch_size = 1 and num_workers=0 as shown below, otherwise it gives me the above mentioned error:
<img width="593" alt="Screen Shot 2021-03-29 at 6 47 52 PM" src="https://user-images.githubusercontent.com/72773297/112826425-8d608b80-90bf-11eb-98de-38cd67d601c8.png">
A learning curve as follows, the model is overfitted:
<img width="562" alt="Screen Shot 2021-03-29 at 6 48 26 PM" src="https://user-images.githubusercontent.com/72773297/112826663-e0d2d980-90bf-11eb-9276-53e7249155b9.png">
Do you know what preprocessing is recommended for preventing overfitting in this case? How can I check which part leads to overfitting in this MRI image dataset? Sorry, I am new to this area, that's why I might miss some important preprocessing. 
I applied a transformation as same as the tutorial.
<img width="594" alt="Screen Shot 2021-03-29 at 7 01 44 PM" src="https://user-images.githubusercontent.com/72773297/112827642-33f95c00-90c1-11eb-9916-ba061eb156ae.png">

**Comment by mattwarkentin** [comment]
Hi @fepegar,

I just wanted to jump into this thread instead of creating an entirely new one. Do you think there should be some sort of way for the user to configure the tolerance for comparing the consistency of some of the attributes, such as the spacing/affine? I have found that rounding errors occur when some files are converted between types (DICOM <-> NRRD, for example), but they should have almost no practical effect.

Of course, in these cases I can simply resample the mask to the image space before, let's say, resampling them both to isotropy or doing some other transform. But this means I am having to resample twice in my preprocessing pipeline rather than just once (to isotropy), which I would prefer to avoid.

What are your thoughts? PyRadiomics uses `1e-16` by default, but I have had Slicer/pyrad devs tell me this tolerance may be too strict so I have often had to override this. Maybe the user should be able to pass a tolerance to `torchio.Subject()` and, if present, this will get used instead of the internal tolerance levels. Thoughts?

https://github.com/fepegar/torchio/blob/6533caef35a5fbaf26197ddf59df319a2eaa3678/torchio/data/subject.py#L237-L255

**Reply by fepegar** [reply]
This issue is a bit complicated. I suppose a good way to handle this is using SimpleITK to check the similarity, but I don't know if they expose some tools for this. Simply comparing the affines is probably not the best way to do this. If you can think of a nice way of comparing, we could try adding the tolerance kwarg to `Subject`.

That being said, can you maybe use two resample transforms? Each image will be resampled only once:

```python
resample = tio.Compose([
    tio.Resample(include=['t1']),
    tio.Resample(target='t1', exclude=['t1']),
])
```

**Reply by mattwarkentin** [reply]
Thanks for the response. So just so I make sure I understand your suggestion, you are resampling `t1` to isotropy first, then resampling the mask to `t1` and excluding `t1` from this resample. So two total resample tasks occur, rather than 3 if you resample the mask to the image, and then both to isotropy, correct?

**Reply by fepegar** [reply]
Exactly!

**Reply by romainVala** [reply]
Hi there,
I continue this thread, because I do also encounter the problem with small differences in the image affine, 
those difference are already in the files (I do not know why but I guess there are numerical issues with affine storage in the nifti)

Any way in my case I am sure the 2 voxel cube does match, so I would like to overigth the affine

Your proposed solution to reslice, seems to me like an overreaction.
It does the job, but it takes time, and it may add interpolation (may be not if I specify nearest-neighbour)

So can we think to an other solution ?

Either adding a transform to copy affine between subject volumes 
Or adding an option to force getting the default "t1" affine when adding a new volume to a subject (which already have a t1 volume)

may be the first one, is more explicit

**Reply by mattwarkentin** [reply]
I too would be interested in a simpler solution to very slight differences in the image affine (which in my case are due to differences in precision between DICOM and NRRD, and not real differences) instead of having to resample. 

> Either adding a transform to copy affine between subject volumes
I like this option I think. A transform where select images/masks are coerced to use the image affine of a chosen image might be simplest.

**Reply by romainVala** [reply]
i think a preprocessing transform should be easy to write,
let see what @fepegar say

**Reply by fepegar** [reply]
Which errors are you guys getting? Could you please paste the full traceback here?

**Reply by mattwarkentin** [reply]
I won't speak for @romainVala, but for me, my images (`ScalarImage()`) are usually in DICOM directories, and my masks (`LabelMap()`) are usually NRRD format. The spacings ARE the exact same, but due to differences in how they are stored, they differ by some small amount of rounding error.

Trying to perform transforms on these will throw errors due to different spacings. Currently my options are to resample the mask to the image space (or vice versa). But since I know they should have the exact same spacing, it would be less costly to just force the mask to adopt the affine matrix of the image.

**Reply by fepegar** [reply]
Thanks, @mattwarkentin.

> Trying to perform transforms on these will throw errors due to different spacings

Could you please provide a pair of images and some code to reproduce the error?

**Comment by romainVala** [comment]
Hi
I can send you images if really needed, but it does not matter, as soon as you have 2 images with different qform 
you can create on as you pointed out here https://github.com/fepegar/torchio/issues/567#issuecomment-859583694

About the code, Simple transform does not care about different affine : 
for instance
```python
i1 = tio.LabelMap('mask.nii')
i2  = tio.ScalarImage('T1.nii')
suj = tio.Subject({'t1':i2, 'mask':i1})
t = tio.CropOrPad([32,32,32])
```
in this case

`
suj_crop = t(suj)
`

just works fine for me. (any other tio transform too)
Actually only the call to check_consistent_affine is made only in tio.WeightedSampler

it must then be that @mattwarkentin  also use a WeightedSampler


```python
tio.data.WeightedSampler(64,'mask')
for path in tw(suj):
    print(patch['index_ini'])
```

does give me an error

```python

/data/romain/toolbox_python/torchio/torchio/data/sampler/weighted.py in __call__(self, subject, num_patches)
     62             num_patches: Optional[int] = None,
     63             ) -> Generator[Subject, None, None]:
---> 64         subject.check_consistent_space()
     65         if np.any(self.patch_size > subject.spatial_shape):
     66             message = (

/data/romain/toolbox_python/torchio/torchio/data/subject.py in check_consistent_space(self)
    264     def check_consistent_space(self):
    265         self.check_consistent_spatial_shape()
--> 266         self.check_consistent_affine()
    267 
    268     def get_images_dict(

/data/romain/toolbox_python/torchio/torchio/data/subject.py in check_consistent_affine(self)
    260                     f'\n{pprint.pformat(image.affine)}'
    261                 )
--> 262                 raise RuntimeError(message)
    263 
    264     def check_consistent_space(self):

RuntimeError: Images "t1" and "mask" do not occupy the same physical space.

```

**Reply by fepegar** [reply]
Thanks, Romain, that's helpful. An option is to remove the constraint that both images need the same affine. That will leave that responsibility to the user but will avoid getting errors due to tiny differences. The change in the code would be replacing `subject.check_consistent_space()` with `subject.check_consistent_spatial_shape()`. What do you guys think?

**Reply by mattwarkentin** [reply]
> it must then be that @mattwarkentin also use a WeightedSampler

I have never used the WeightedSampler. Spacing errors get thrown for other Transformations (I think it was `Resample` in my case.

**Reply by romainVala** [reply]
that is weird, 
I do not see any other call to check_consistent_spatial_affine, except for the WeightedSampler
@mattwarkentin  could you post a minimal code example with the exact error message you get ?

**Reply by mattwarkentin** [reply]
I was mistaken, the resampling works even with different spacings for the image and mask. However, if there is a very small difference in the affine spacing (due to, let's say, rounding error from different storage formats), this may cause an "off-by-one" issue with the new spatial shape after resampling. This will cause errors in downstream transforms, for example, `tio.CropOrPad()` or `tio.Mask()`.

```py
import numpy as np
import torch
import torchio as tio

img = torch.randn(1, 512, 512, 100)
mask = torch.tensor(img > 0).type(torch.int16)

af1 = np.array([[0.8, 0, 0, 0],
                [0, 0.8, 0, 0],
                [0, 0, 2.50000000000001, 0],
                [0, 0, 0, 1]])

af2 = np.array([[0.8, 0, 0, 0],
                [0, 0.8, 0, 0],
                [0, 0, 2.5, 0],
                [0, 0, 0, 1]])

sub = tio.Subject(
  image = tio.ScalarImage(tensor = img, affine = af1),
  mask = tio.LabelMap(tensor = mask, affine = af2)
)

sub.spacing # throws error

rs = tio.Resample() # Resample to isotropy
sub_rs = rs(sub)

sub_rs.spacing # no longer an error

sub_rs.spatial_shape # throws error

tio.CropOrPad(50)(sub_rs) # throws error

tio.Mask(masking_method = "mask")(sub_rs)  # throws error
```

**Reply by romainVala** [reply]
ok, thanks for the detail example, 
indeed, small difference (in voxel resolution) does have some consequence
at least the fernando's proposition to replace your
` tio.Resample() `
with
```python
resample = tio.Compose([
    tio.Resample(include=['image']),
    tio.Resample(target='image', exclude=['image']),
])
```

make the job

> Thanks, Romain, that's helpful. An option is to remove the constraint that both images need the same affine. That will leave that responsibility to the user but will avoid getting errors due to tiny differences. The change in the code would be replacing `subject.check_consistent_space()` with `subject.check_consistent_spatial_shape()`. What do you guys think?

given this example, this will not be enough

Either we go to
1_  allowing overight of affin in the image __init__
2_ or adding a transform to copy affine, spacing ect ...

1_ would be simpler no ?

**Reply by romainVala** [reply]
Actually a transform CopyAffine, is also easy to do, and easy to insert in existing code,
we have a local version, 
@fepegar  should we propose a PR ?

**Reply by romainVala** [reply]
@fepegar any preference between the 2 options  ?

**Reply by fepegar** [reply]
Option 1 is not practical because if you try to read the `affine` property of an image, the data will be loaded to RAM and that might not be desired. We could modify the behavior to read the affine using NiBabel without loading the data, but that's a different topic. I suppose a `CopyAffine` transform would be useful!

**Comment by romainVala** [comment]
I am surprise this check is not needed elsewhere, but ok, one solution would be to remove it, 
But when different affine are in place this is often a sign that the user must double check what is going on. 
So I thing to give an error here is a good thing, though it is annoying for tiny differences, but coding is often anoying ...

So I would prefer, to add an argument to WeigtedSampler, to explicitly ask to not do the check ...
```python
tio.data.WeightedSampler(64,'mask', check_affine_consitency=False)
```


May be an other solution (instead of creating a new transform, that make an affine copy) would be to specify it at the creation 
something like that

```python
i2  = tio.ScalarImage('T1.nii')
i1 = tio.LabelMap('mask.nii', affine = i2.affine)
suj = tio.Subject({'t1':i2, 'mask':i1})
```
but the affine argument seems not to be taken into account 

```python
In [73]: np.max(np.abs(suj.t1.affine - suj.mask.affine))
Out[73]: 0.00064480513012474

```
I thing this argument was only though for tensor input ...

**Comment by ryancinsight** [comment]
@fepegar i’m getting this error trying to implement a cGan on T1 and T2 image sets. However, Croporpad does not seem to be working when subject dataset has multiple scaled image types. Is this possible or can you only have one scalar image and one mask?

**Reply by mattwarkentin** [reply]
Hmm, I tried looking into this, I wonder if this is a bug @fepegar. Should `tio.CropOrPad()` work on a subject with scalar images that possibly have different spacings and spatial shapes? It seems like it should, unless I am missing something.

Here is some example code to play with:
```py
import torchio as tio

# These `ScalarImage`s have different spacings and orientations
brain = tio.datasets.Colin27().t1
ct = tio.datasets.Slicer('CTChest').CT_chest

sub = tio.Subject(ct=ct, brain=brain)

# This raises an error
tio.CropOrPad(100)(sub)
```
```
RuntimeError: More than one spatial_shape found in subject images:
{'brain': (181, 217, 181), 'ct': (512, 512, 139)}
```
```py
# Transform images in same orientation and spacing
transforms = tio.Compose([
    tio.ToCanonical(),
    tio.Resample(1)
])

sub_t = transforms(sub)

# Still errors
tio.CropOrPad(100)(sub_t)
```
```
RuntimeError: More than one spatial_shape found in subject images:
{'brain': (181, 217, 181), 'ct': (391, 391, 348)}
```

**Reply by ryancinsight** [reply]
I get the same error with EnsureShapeMultiple as well. I'm not sure I understand why the check is at the beginning. I can set T1=T1 and T2=T1 so that they are the same image and these seem to work. Thus, the only way I've been able to bypass is create subjects zip style and apply the croporpad individually.

`transform = tio.Compose([
    tio.ToCanonical(),
    tio.Resample((2,2,2)),
    tio.CropOrPad((100,100,100)),
])

subjects = [tio.Subject(
        T1=transform(tio.ScalarImage(T1_path)),
        T2=transform(tio.ScalarImage(T2_path)),
        PD=transform(tio.ScalarImage(PD_path)),
    ) for (T2_path, T1_path, PD_path) in tqdm(zip(T2_paths, T1_paths, PD_paths))]
dataset = tio.SubjectsDataset(subjects)`

**Comment by romainVala** [comment]
Well the way it is implemented, you need to have the same spatial shape if you have multiple images in your subject
(because it compute one pad and/or one crop from the images shape compare to the request end shape.)

but I see your point it make sense to have different shape (that is why you add this croporpad transform, to have the same shape at the end)

It is possible to change the code : adding a loop over each subject image, to perform a specific crop and / or pad. (like it is done in the resample transform)

**Comment by fepegar** [comment]
Hmm I think this is a bit delicate. If the images have different shapes, chances are they are not in the same physical space, so their orientation and spacing might be different. Using `CropOrPad` would probably produce undesirable results. For those cases, `Resample` should be used instead.

@ryancinsight, could you please share a pair of your images so we understand the issue a bit better?

**Reply by ryancinsight** [reply]
Actually, I'm just using the IXIdataset and bringing in the T1 and T2 images for the patient brain. The goal is image synthesis from T1 to T2 using cGan so that's why I need to croporpad after the resampling to 1mm.

**Reply by ryancinsight** [reply]
I see your point though, my assumption is the images are registered and the affine transforms are applied. I may need a regpadandcrop that uses something like elastix to register and resample to the target image affine.

**Reply by fepegar** [reply]
> I see your point though, my assumption is the images are registered and the affine transforms are applied.

Exactly, and that's not the case with the raw images.

I suggest you register the T2s to the T1s using rigid registration and the T1s to the MNI space using affine registration ([example with ANTS here](https://github.com/fepegar/resseg#episurg)) and. Then, resample everything on the MNI space and your life will be better.

**Reply by ryancinsight** [reply]
Impressive work, thank you for the example! Unfortunately, I have windows and have had nightmares compiling Antspy so I'll retool it with simpleelastix.

Best regards!

**Reply by romainVala** [reply]
> Hmm I think this is a bit delicate. If the images have different shapes, chances are they are not in the same physical space, so their orientation and spacing might be different. Using `CropOrPad` would probably produce undesirable results. For those cases, `Resample` should be used instead.

not so sure, the fact that the images are coregister (ie in the same physical space) is a uncorrelated issue (you can have mismatch with same image having the same shape ...)

So the mismatch in shape is simply due here to the different FOV (field of view) between T1 and T2 in @ryancinsight 's case. As the resample (does not use a target space, then both image keep their native FOV).
so a simple solution is then to use

```python 
tio.Resample((2,2,2), target='T1' )
```
I can then imagine case, where the T1 and T2 image have the same resolution but different FOV, so if one may want to avoid resample, one will need a `CropOrPad` with different crop or pad to get a common target shape. 

but ok, you may be rigth, (after trying to show the opposit) ... : It will produce undesirable results, if the FOV center are different. And if one want to use "mask_name" argument, to have a consitent target FOV center, one will need 2 mask (because of the 2 resolutions) so

So no need to change the code, also because an alternative is to use 2 CropOrPad, 

```python 
tio.Compose([
tio.CropOrPad(target=100, include='T1') ,
tio.CropOrPad(target=100, include='T2') ,
])

```

**Reply by fepegar** [reply]
@ryancinsight FYI I've had good results on Windows using NiftyReg. I also typically segment the brain using ROBEX and use the masks to aid the registration process.

@romainVala yes I meant `Resample` using one of the images as the target. Yesterday, a colleague asked me for help because he had a label map in the same space as the scalar image, but with different FOVs. Therefore, both images had the same shape after `CropOrPad` but were not aligned in the physical space. I explained to him how to use `Resample` for this and he was happy. I'm thinking of writing an FAQ section in the docs, this issue should be there.

**Reply by mattwarkentin** [reply]
@fepegar I think it would be nice to have an FAQ or "common issues" section to provide small explanations and code snippets for these sorts of examples. It might not be immediately obvious (it wasn't to me) that `tio.Resample(target='name')` doesn't just resample all the images to the spacing of `name`, but also to have the same spatial shape.

**Reply by fepegar** [reply]
We agree then :)

**Reply by romainVala** [reply]
one should also emphasis that this Resample transform add an extra interpolation. then the solution with 2 CropOrPad, is still worth to mention ... I think ...

**Reply by ryancinsight** [reply]
I didn't realize this was possible, so if already registered I can resample all to the same image space with target='name' ?

**Reply by mattwarkentin** [reply]
@ryancinsight Yes, exactly!
